### PR TITLE
RWAHS-309 Sort

### PR DESCRIPTION
**Pull request by leftclickben**
Adds sort functionality, which completes this task (or is it an epic?)

### CODE REVIEW

File: public/app/models/ListModeSwitcher.js [hunk 1]
```diff
@@ -0,0 +1,25 @@
+(function () {
```
Comment (leftclickben) [inline public/app/models/ListModeSwitcher.js]: This module was just pulled out from `SearchPage` for code structure consistency (with `ListSorter` and `ListPager`). Not much to see here.

File: public/app/models/ListSorter.js [hunk 2]
```diff
@@ -0,0 +1,49 @@
+(function () {
```
Comment (leftclickben) [inline public/app/models/ListSorter.js]: This is the meat of the change.

File: public/app/models/ListSorter.js [hunk 3]
```diff
@@ -0,0 +1,49 @@
...
+
+                this.sortedList = ko.pureComputed(function () {
+                    var data = sourceObservable(),
+                        field = this.field(),
+                        fieldDefinition = _.find(fieldsObservable(), { key: field }),
+                        direction = this.direction();
+                    if (!field || !direction) {
+                        return data;
+                    }
+                    return _.orderBy(
```
Comment (leftclickben) [inline public/app/models/ListSorter.js]: https://lodash.com/docs#orderBy

File: public/app/models/ListSorter.js [hunk 4]
```diff
@@ -0,0 +1,49 @@
...
+                                // Field-specific value translation.
+                                value = fieldDefinition.sort(value);
+                            }
+                            if (_.isString(value)) {
+                                // Case-insensitive sorting of string values.
+                                value = value.trim().toLowerCase();
+                            }
+                            if (!value) {
+                                // Use sentinel values to push empty values to the end, for both sort directions.
+                                value = (direction === 'asc' ? '~' : '!');
```
Comment (leftclickben) [inline public/app/models/ListSorter.js]: http://www.asciitable.com/

**Comment by leftclickben** [comment]
Raised in backlog:

https://youtrack.gaiaresources.com.au/youtrack/issue/RWAHS-407
To improve the sort for accession numbers.

https://youtrack.gaiaresources.com.au/youtrack/issue/RWAHS-408
To improve the sort for date fields.

**Comment by kehh** [comment]
Looks very nice, but doesn't appear to sort

**Comment by leftclickben** [comment]
Works in firefox, not in chrome.  I know the issue too, I did something temporarily in the bindings that wasn't quite right.  Will fix, but in the meantime if you use firefox you should be able to continue the review.

File: public/app/ui/components/controls/ListControlsComponent.js [hunk 5]
```diff
@@ -15,22 +15,63 @@
...
                 }
 
-                this.availableResultsModes = ko.pureComputed(function () {
+                this.availableSortFields = ko.pureComputed(function () {
+                    return _.flatten(_.map(
+                        parameters.sorter.availableSortFields(),
+                        function (field) {
+                            return [
+                                {
+                                    label: field.labelText + ' asc',
```
Comment (kehh) [inline public/app/ui/components/controls/ListControlsComponent.js]: `asc` and `desc` are not necessarily 'in common parlance', at least label them as `ascending` and `descending`. I know space is an issue, but this is the public frontend so user training isn't going to cover it.

Comment (kehh) [inline public/app/ui/components/controls/ListControlsComponent.js]: or add longLabel to the `option`'s title as well as the `aria-label`.

File: public/app/ui/components/controls/list-controls.html [hunk 6]
```diff
@@ -1,9 +1,20 @@
 <div class="list-controls-container">
     <div>
         <div class="input-group input-group-sm">
+            <span class="input-group-addon">Sort:</span>
+            <select class="form-control">
+                <option value="" aria-label="Display the results in natural order" data-bind="click: resetSort">None</option>
+                <!-- ko foreach: availableSortFields -->
+                <option data-bind="text: label, attr: { value: value, selected: active, 'aria-label': longLabel }, click: click"></option>
```
Comment (kehh) [inline public/app/ui/components/controls/list-controls.html]: I guess this is where you'd add `,title: longLabel`

**Comment by kehh** [comment]
Just the comments about usability of the sort labels. Also found this https://youtrack.gaiaresources.com.au/youtrack/issue/RWAHS-410, but that's already in `master`.

**Pull request merged**